### ISSUE TITLE
Questions about design goals, roadmap, etc.

### ISSUE THREAD

**Issue by mkaze** [JetBrains/KotlinDL#69]
I am currently learning Kotlin and in order to enhance my learning experience I was looking for a project accepting contributions. So I am excited to find this library, especially considering the fact that I have a relatively fair amount of experience working with Keras/TF in Python. However, I have a few questions about some of the high-level aspects of the project and I would appreciate if the maintainers could answer them:

- Is the goal to mimic the Python API of Keras as close as possible, or instead, should it focus on "what's the idiomatic way to implement XYZ feature in Kotlin?" regardless of how the that feature has been implemented in Python API of Keras (i.e. kotlinic vs. pythonic)? As I see, currently there is a rather strong focus to follow Keras API.
- What's the plan for switching to TF 2+ Java APIs?
- Is TF the only backend to be supported, or is there a plan to add support for more backends (e.g. PyTorch)?
- How is this going to differ from Deeplearning4J?
- Is there a long-term goal to get rid of the backend(s) in future and replace them with Kotlin implementations and libraries (i.e. similar to what DL4J has done, though from the beginning)?

Thanks!

**Comment by zaleslaw** [comment]
Great questions, I'll try to answer all of them.

> I am currently learning Kotlin and in order to enhance my learning experience I was looking for a project accepting contributions. So I am excited to find this library, especially considering the fact that I have a relatively fair amount of experience working with Keras/TF in Python. However, I have a few questions about some of the high-level aspects of the project and I would appreciate if the maintainers could answer them:
> 
> * Is the goal to mimic the Python API of Keras as close as possible, or instead, should it focus on "what's the idiomatic way to implement XYZ feature in Kotlin?" regardless of how the that feature has been implemented in Python API of Keras (i.e. kotlinic vs. pythonic)? As I see, currently there is a rather strong focus to follow Keras API.

The project is started as a copy of Keras at the top of TensorFlow and Keras is the most popular high-level framework in DL at this moment (top-level frameworks based on PyTorch are not so popular)

Also, it supports the transfer learning on Keras models and we need to be fully compatible with it (has the same layers with the same inputs and outputs, and so on). 

But we started the experiment with KotlinDSL for preprocessing and are going to design an alternative API for the building NN as a graph (I suppose it will be an open design doc and I'll share it here, in this issue and discussion chapter.

But it will be not limited to Keras API only, we think it will be extended to have customizable parts and training cycle like in PyTorch

> * What's the plan for switching to TF 2+ Java APIs?

TF Java API 2.x is a low-level API under development. The team is great and project too, but.

It has no advantages at this moment for switching:
- not so many new ops
- no performance speedup
- possible memory leak
- unstable and experimental API
- we could load TF 2.x models written in Keras

Resume: we switch on TF Java API 2.x if it will be more stable and a few critical issues will be solved

> * Is TF the only backend to be supported, or is there a plan to add support for more backends (e.g. PyTorch)?

Fully answered [here](https://github.com/JetBrains/KotlinDL/discussions/49)

> * How is this going to differ from Deeplearning4J?

DL4j is a great project, and we have no goal to repeat it or rewrite it from scratch. 
Correct me if I'm wrong about the idea of DL4j to have its own runtime and integrations with other DL frameworks.
We are not going to write our own runtime, no chance to do it better than TF or Torch teams

Our idea to cover main and popular cases for production, data engineers and be a good entry point for students.
This is well described in the presentation ["Deep Learning with KotlinDL"](https://youtu.be/jCFZc97_XQU?t=291)

> * Is there a long-term goal to get rid of the backend(s) in future and replace them with Kotlin implementations and libraries (i.e. similar to what DL4J has done, though from the beginning)?

There is a few NumPy-like libraries in Kotlin ecosystem: [multik ](https://github.com/Kotlin/multik) and [viktor](https://github.com/JetBrains-Research/viktor), 

We are going to use Multik to build inference or maybe transfer learning in Kotlin for mobile devices but as I mentioned earlier we have no goal to build and maintain our own Tensor library from scratch, with high probability it will be wrappers for TF or PyTorch

> Thanks!
Thank you for your questions

So, if you have any ideas about Kotlin idiomatic stuff related to NN or data preprocessing, I invite you to collaborate here or in dicsussion chapter

**Comment by mkaze** [comment]
@zaleslaw Thanks a lot for your comments. That makes it clearer now.
One more question: I am not familiar with TF Java API, so do you see any advantage in writing a Kotlin API for TF (probably as a separate project and a long-term goal)?

> So, if you have any ideas about Kotlin idiomatic stuff related to NN or data preprocessing, I invite you to collaborate here or in the discussion chapter.

Sure! Although, as I mentioned earlier, I am new to Kotlin and I am still learning it and studying the codebase to understand what's going on. Anyways, I'll try to work on some of the "good first issues" and also share my ideas. Thanks again!

**Comment by zaleslaw** [comment]
Currently, Kotlin API for low-level operands have a chance to become a part of tensorflow/java project (it has a draft PR), if it will be mature and useful, we will not create yet one, if it will be frozen or unfinished, so, Kotlin API for low-level TF should be created too (maybe as a separate module in this project)
